### PR TITLE
Feature/update slack status

### PR DESCRIPTION
**Pull request by TNagaraja**
Added feature - user inputs a message including an emoji and status message to update their status. Inputting /wfh clear will reset status. A message consisting of just an emoji will by set a default status message of 'Working from home'.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -3,16 +3,16 @@
   "version": "1.0.0",
   "description": "Tell Slack you're working from home today 🏡",
   "repository": {
-	  "type": "git",
-	  "url": "https://github.com/Symphono/wfh-slack-command"
```
Comment (jspringer2020) [inline package.json]: Make sure you have the proper `spaces` vs `tabs` configured and revert this change
- [ ] revert change

File: src/modify-status.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
+var slackApi = require('./slack-api');
+var emoji = require('node-emoji');
+
+var parseText = function (text) {
```
Comment (jspringer2020) [inline src/modify-status.js]: This method looks like its returning a count and not so much a `parse`. What is the use of this function and let's make sure we name it accordingly.
- [ ] rename function

File: src/modify-status.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+			statusText = text.substring(unicodeCount);
+		}
+		if (statusText === '')
+		{
+			statusText = 'Working from home';
+		}
+	}
+	else
+	{
+		// Status Emoji not specified at beginning of text or no emoji at all, return empty strings
```
Comment (jspringer2020) [inline src/modify-status.js]: What about the use case where I don't have an emoji but I still pass a a status to be updated?

File: src/modify-status.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	}
+	return { statusEmoji: statusEmoji, statusText: statusText };
+}
+
+module.exports = (text, token) => {
+	var encodedJSON = 'token='.concat(token, '&profile=', encodeURIComponent('{"status_text":"'.concat('', '","status_emoji":"', '', '"}')));
+	if (text !== 'clear' && text !== '')
```
Comment (jspringer2020) [inline src/modify-status.js]: If the user passes `clear` we should be removing the user's status as we are assuming they previously had a status as WFH. We also need to have logic around no message and defaulting the icon and status message.
- [ ] Add logic for clearing status
- [ ] Wire in logic so when no status is passed, we still set a default status and icon

File: test/modify-status.spec.js [hunk 5]
```diff
@@ -0,0 +1,119 @@
+var Chance = require('chance');
```
Comment (jspringer2020) [inline test/modify-status.spec.js]: FYI, you can also do `var chance = require('chance')();` notice I executed the function that comes back from `require` as opposed to creating the variable on line 9.

File: test/modify-status.spec.js [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+		};
+		modifyStatus.__set__({
+			slackApi: fakeSlackApi
+		});
+	}
+
+	describe('When processing a request', () => {
+		beforeEach(() => {
+			token = chance.string();
+			text = 'some status message';
```
Comment (jspringer2020) [inline test/modify-status.spec.js]: Let's use `chance` to generate us some text as well.
- [ ] Use `chance` to generate text

File: test/modify-status.spec.js [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+						);
+					});
+		});
+		describe('And the user passes in an empty text', () => {
+			beforeEach(() => {
+				text = '';
+				return act();
+			});
+			it('should send back to slack an encoded URI in which the profile parameter is set to a payload of empty strings for statusEmoji and statusText', () =>
+				sinon.assert.calledWith(fakeSlackApi.sendResponse, 'slack.com/api/users.profile.set', 'token='.concat(token, '&profile=', encodeURIComponent('{"status_text":"'.concat('', '","status_emoji":"', '', '"}'))))
```
Comment (jspringer2020) [inline test/modify-status.spec.js]: If the user hasn't passed any text, what is the expected behaviour? That expected behaviour should be the `it`. Remember, the goal of business driven testing is readibility. Will a user know what `encoded URI in which the profile parameter...` means? Or will they know `should clear the user's status` or `should set the user's status from the supplied text`?

File: test/modify-status.spec.js [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+					describe('And the text contains the status message before the emoji', () => {
+						beforeEach(() => {
+							text = text.concat(emoji.random().emoji);
+							return act();
+						});
+						it('should send back to slack an encoded URI in which the profile parameter is set to a payload of empty strings for statusEmoji and statusText', () =>
+							sinon.assert.calledWith(fakeSlackApi.sendResponse, 'slack.com/api/users.profile.set', 'token='.concat(token, '&profile=', encodeURIComponent('{"status_text":"'.concat('', '","status_emoji":"', '', '"}'))))
+						);
+					});
+					describe('And the user passes in \'clear\'', () => {
```
Comment (jspringer2020) [inline test/modify-status.spec.js]: What's more readable, the current text or `And the user cleared the WFH`?

File: test/modify-status.spec.js [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+					describe('And the text contains multiple emojis', () => {
+						beforeEach(() => {
+							text = randomEmoji.concat(randomEmoji, text);
+							return act();
+						});
+						it('should send back to slack an encoded URI in which the profile parameter is set to a payload of statusEmoji and statusText, where statusText contains the emojis after the first', () =>
+							sinon.assert.calledWith(fakeSlackApi.sendResponse, 'slack.com/api/users.profile.set', 'token='.concat(token, '&profile=', encodeURIComponent('{"status_text":"'.concat(randomEmoji, 'some status message', '","status_emoji":":', emoji.which(randomEmoji), ':"}'))))
+						);
+					});
+					describe('And the text contains the status message before the emoji', () => {
```
Comment (jspringer2020) [inline test/modify-status.spec.js]: This is good, it describes what we are doing. Only change I'd recommend is instead of `and the text` I would say, 'And the WFH contains`

File: test/modify-status.spec.js [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+							sinon.assert.calledWith(fakeSlackApi.sendResponse, 'slack.com/api/users.profile.set', 'token='.concat(token, '&profile=', encodeURIComponent('{"status_text":"'.concat('some status message', '","status_emoji":":', emoji.which(randomEmoji), ':"}')))));
+
+					});
+					describe('And the text contains an emoji followed immediately by a status message', () => {
+						beforeEach(() => {
+							text = randomEmoji.concat(text);
+							return act();
+						});
+						it('should send back to slack an encoded URI in which the profile parameter is set to a payload of statusEmoji and statusText', () =>
+							sinon.assert.calledWith(fakeSlackApi.sendResponse, 'slack.com/api/users.profile.set', 'token='.concat(token, '&profile=', encodeURIComponent('{"status_text":"'.concat('some status message', '","status_emoji":":', emoji.which(randomEmoji), ':"}'))))
```
Comment (jspringer2020) [inline test/modify-status.spec.js]: FYI, I'm like 90% sure, you can write this as `fakeSlackApi.sendResponse.calledWith(...)`

**Review by jspringer2020 — CHANGES_REQUESTED** [review]
Need to address a few things.

File: package.json [hunk 11]
```diff
@@ -3,8 +3,8 @@
   "version": "1.0.0",
   "description": "Tell Slack you're working from home today 🏡",
   "repository": {
-    "type": "git",
-    "url": "https://github.com/Symphono/wfh-slack-command"
+	"type": "git",
+	"url": "https://github.com/Symphono/wfh-slack-command"
```
Comment (jspringer2020) [inline package.json]: Looks like we still have a spacing issue here
- [ ] resolve spacing issue

File: src/modify-status.js [hunk 12]
```diff
@@ -1,60 +1,60 @@
 var slackApi = require('./slack-api');
 var emoji = require('node-emoji');
 
-var parseText = function (text) {
+var status = {
+    status_text: 'Working from Home',
+    status_emoji: ':house_with_garden:'
+	}
```
Comment (jspringer2020) [inline src/modify-status.js]: always end lines with a semi colon..I'm actually shocked the linter didn't catch this...

- [ ] add semi colon

File: src/modify-status.js [hunk 13]
```diff
@@ -1,60 +1,60 @@
...
+
+var defaultStatus = function () {
+		status.status_text = 'Working from home';
+		status.status_emoji = ':house_with_garden:';
+}
 
-	var unicodeCount = 0;
+var getEmojiLength = function (text) {
```
Comment (jspringer2020) [inline src/modify-status.js]: 👍  much better naming

File: src/modify-status.js [hunk 14]
```diff
@@ -1,60 +1,60 @@
...
+var setStatusEmojiAndText = function (text) {
+	if (text === 'clear')
 	{
-		statusEmoji = text.substr(0, unicodeCount);
-		if (text.substr(unicodeCount, 1) === ' ')
-		{
-			statusText = text.substring(unicodeCount + 1);
-		}
-		else
+		status.status_text = '';
```
Comment (jspringer2020) [inline src/modify-status.js]: where are we creating `status`? if we aren't passing it in its likely globally scoped which is like a static instance.

File: src/modify-status.js [hunk 15]
```diff
@@ -1,60 +1,60 @@
...
-		var unicodeCount = parseText(text);
-		var statusEmoji = getStatusEmojiAndText(text, unicodeCount).statusEmoji;
-		var statusText = getStatusEmojiAndText(text, unicodeCount).statusText;
-		if (statusEmoji !== '')
-		{
-			encodedJSON = 'token='.concat(token, '&profile=', encodeURIComponent('{"status_text":"'.concat(statusText, '","status_emoji":":', emoji.which(statusEmoji), ':"}')));
-		}
-	}
+	setStatusEmojiAndText(text);
+	var encodedJSON = 'token='.concat(token, '&profile=', encodeURIComponent(JSON.stringify(status)));
```
Comment (jspringer2020) [inline src/modify-status.js]: I see what you are doing now...the `setStatusEmojiAndText()` should return a structure that you then stringify, otherwise, the `status` variable above acts as a static instance. Inside of `setStatusEmojiAndText()` you should call `getDefaultStatus()` which returns you a new instance with default values instead of mutating the static variable. Let me know if you have questions.
- [ ] Update to not use a static variable for storing status.

**Review by jspringer2020 — COMMENTED** [review]
Getting much closer...

**Comment by MarkLeMerise** [comment]
Hey guys, please ignore the Travis CI check for now. I was experimenting on integrating it with this repo and it probably got kicked off with @TejNagaraja's commits. I turned it off for now as there is some additional configuration that needs doing.

**Review by jspringer2020 — CHANGES_REQUESTED** [review]
There is still some more stuff to be tweaked here.

File: test/modify-status.spec.js [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+		beforeEach(() => {
+			token = chance.string();
+			text = chance.sentence();
+			status = {
+				status_text: 'Working from home',
+				status_emoji: ':house_with_garden:'
+			}
+			refreshMocks();
+		});
+		describe('And the user passes in a nonempty text', () => {
```
Comment (jspringer2020) [inline test/modify-status.spec.js]: What is `nonempty text`? Does that contain an emoji? How about `And the user set the status text`. By doing that, we ca now build on it with a nested describe of `And the user set an emoji` so the output would read `... And the user set the status text. And the user set an emoji.` ⬅️  is very human readable.
- [ ] update `describe message`

File: test/modify-status.spec.js [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+					var randomEmoji = emoji.random().emoji;
+					describe('And the WFH is \'🚞 riding a train\' and token is \'super_secret_token\'', () => {
```
Comment (jspringer2020) [inline test/modify-status.spec.js]: I really like where your head is at with the `and token is ...`. You are completely describing the setup of the test, which is fantastic...Since `token` is kind of a pass along value, meaning we don't really derive any logic from it -- any change it value doesn't impact the code, we don't have to `describe` it, we simply need to assert `it` to make sure the value we supplied is what was called in slack...

Also, I really like your descriptiveness of this test's `describe`. One thing to be careful of, and why `chance` is handy is that we aren't really testing specific text or emoji, so we can randomly generate this via `chance` and the emoji library. See my comment above about making this slightly more generic. It will mean you will have to store the emoji and the status text in variables and compare them in the assertion -- which you are already doing.
- [ ] randomly generate text and an emoji

File: test/modify-status.spec.js [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+			token = chance.string();
+			text = chance.sentence();
+			status = {
+				status_text: 'Working from home',
+				status_emoji: ':house_with_garden:'
+			}
+			refreshMocks();
+		});
+		describe('And the user passes in a nonempty text', () => {
+					var randomEmoji = emoji.random().emoji;
```
Comment (jspringer2020) [inline test/modify-status.spec.js]: Generating your `randomEmoji` should go in a `beforeEach()` otherwise you are reusing the same emoji over and over...this is a little gotcha from scoping..the reason this happens is because the way the test runner works it actually executes all the test files so any declarations or code happening outside of a `before()`, `act()` or `it()` will happen when the tests initialize which only happens once...from there the test runner can gather the "execution plan" to execute all the `before()` then execute the `act()` then execute the `it()...so the code would look like
```
var randomEmoji;
beforeEach(() => randomEmoji = emoji.random(););
```
- [ ] ensure a random emoji is generated for every test.

File: test/modify-status.spec.js [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+							status.status_text = randomEmoji.concat(text);
+							status.status_emoji = ':'.concat(emoji.which(randomEmoji), ':');
+							text = randomEmoji.concat(randomEmoji, text);
+							return act();
+						});
+						it('should set status emoji to first emoji and status message to remainder of text', () =>
+							sinon.assert.calledWith(fakeSlackApi.sendResponse, 'slack.com/api/users.profile.set', 'token='.concat(token, '&profile=', encodeURIComponent(JSON.stringify(status))))
+						);
+					});
+					describe('And the WFH contains the status message before the emoji', () => {
```
Comment (jspringer2020) [inline test/modify-status.spec.js]: You really thought about all the different use cases here. That is fantastic. I can't really think of any test cases you have missed. I really like the 2 emoji and emoji at the end of the status test cases. I hadn't even though of those, so that's great. Good work thinking through all the scenarios. 👍

**Review by jspringer2020 — CHANGES_REQUESTED** [review]
Great work here overall. The code is definitely coming along. Just a few tweaks to the tests and I think we can get final approval from @MarkLeMerise

File: src/modify-status.js [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+	{
+		status.status_text = text.substring(messageBegin);
+	}
+	return status;
+}
+
+module.exports = (text, token) => {
+	var status = buildStatusJSON(text);
+	var encodedJSON = 'token='.concat(token, '&profile=', encodeURIComponent(JSON.stringify(status)));
+	return slackApi.sendResponse('slack.com/api/users.profile.set', encodedJSON);
```
Comment (MarkLeMerise) [inline src/modify-status.js]: As a matter of architecture, I would suggest creating a `slackApi.setStatus(encodedJSON)` method instead of overloading `sendResponse`. That way, if the URL for that API endpoint ever changes, you can change the API wrapper but not your calling code.

File: src/modify-status.js [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+
+function buildStatusJSON (text) {
+	if (text === 'clear')
+	{
+		return getBlankStatus();
+	}
+	var status = getDefaultStatus();
+	var messageBegin = getEmojiLength(text);
+	if (isEmojiChar(text.codePointAt(0))) {
+		status.status_emoji = ':'.concat(emoji.which(text.substr(0, messageBegin)), ':');
```
Comment (MarkLeMerise) [inline src/modify-status.js]: Consider using [template strings](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) for string concatenation. While strings are technically array-like, `concat` is usually used for actual arrays. I think there were a couple other places where you could clean up the intent as well.

File: test/modify-status.spec.js [hunk 22]
```diff
@@ -0,0 +1,155 @@
...
+			text = chance.sentence();
+			status = {
+				status_text: 'Working from home',
+				status_emoji: ':house_with_garden:'
+			}
+			refreshMocks();
+		});
+		describe('And the user set a status', () => {
+			describe('And the user also set an emoji', () => {
+				var someToken = chance.string();
```
Comment (MarkLeMerise) [inline test/modify-status.spec.js]: It might be easier to add a `beforeEach` here setting the random values for `text`, `token`, and `emoji`. That way you can eliminate those assignments from all the nested cases. Usually, the `describe` will indicate what work will be done in the next `beforeEach`. For example, `'And the user set a status'` probably means there'd be a `beforeEach` generating a random status.

File: src/modify-status.js (hunk 20 above)
Comment (MarkLeMerise) [inline src/modify-status.js]: Furthermore, you may want to double check that this is sending the request as you want. For example, the payload (`encodedJSON`) is being formatted as `application/x-www-form-urlencoded` via `encodeURIComponent` but if you look at the implementation of `slackApi.sendResponse`, the request is _actually_ being sent as `application/json`. You might get away with not having to URL-encode the payload (sending as straight JSON) _or_ if that doesn't work, when you define your `setStatus` method, you can manipulate the JSON object there into a URL-encoded string ([query-string](https://github.com/sindresorhus/query-string) is a good library for this).

**Review by MarkLeMerise — CHANGES_REQUESTED** [review]
Overall looks solid, but there were a few points for potential cleanup. Also, is there a commit coming which calls this function in the main workflow?

**Comment by TNagaraja** [comment]
Passed tests against slack api

**Review by MarkLeMerise — APPROVED** [review]

**Review by jspringer2020 — COMMENTED** [review]
Merge away
